devito/passes/clusters/misc.py

```diff
@@ -304,21 +296,35 @@ def choose_element(queue, scheduled):
             if not scheduled:
                 return queue.pop()
 
-            # Heuristic: let `k0` be the key of the last scheduled node; then out of
-            # the possible schedulable nodes we pick the one with key `k1` such that
-            # `max_i : k0[:i] == k1[:i]` (i.e., the one with "the most similar key")
-            key = self._key(scheduled[-1])
-            for i in reversed(range(len(key) + 1)):
-                candidates = [e for e in queue if self._key(e)[:i] == key[:i]]
+            k = self._key(scheduled[-1])
+            m = {i: self._key(i) for i in queue}
+
+            # Process the `strict` part of the key
+            candidates = [i for i in queue if m[i].itintervals == k.itintervals]
+
+            if k.guards:
+                compatible = [i for i in candidates if m[i].guards == k.guards]
+                candidates = compatible or candidates
+
+            if k.syncs:
+                compatible = [i for i in candidates if m[i].syncs == k.syncs]
+                candidates = compatible or candidates
+
+            # Process the `weak` part of the key
+            for i in reversed(range(len(k.weak) + 1)):
+                choosable = [e for e in candidates if m[e].weak[:i] == k.weak[:i]]
```

Comment:
I dont like this var name but ok

Reply:
the choosable among the candidates -- I think it's OK! why don't you like it? Im open to alternatives!

Reply:
hm, not sure....maybe like `filtered`, `valid`, `eligible` ?